libmeep_geom/meep_geom.cpp

```diff
@@ -0,0 +1,1540 @@
+{
+  if (dims == CYLINDRICAL) {
+    dimensions = 2;
+    dim = meep::Dcyl;
+  }
+  else {
+    dimensions = dims;
+    dim = meep::ndim(dims - 1);
+  }
+}
+
+vector3 vec_to_vector3(const meep::vec &pt)
+{
+  vector3 v3;
+  
+  switch (pt.dim) {
+  case meep::D1:
+    v3.x = 0;
+    v3.y = 0;
+    v3.z = pt.z();
+    break;
+  case meep::D2:
+    v3.x = pt.x();
+    v3.y = pt.y();
+    v3.z = 0;
+    break;
+  case meep::D3:
+    v3.x = pt.x();
+    v3.y = pt.y();
+    v3.z = pt.z();
+    break;
+  case meep::Dcyl:
+    v3.x = pt.r();
+    v3.y = 0;
+    v3.z = pt.z();
+    break;
+  }
+  return v3;
+}
+
+meep::vec vector3_to_vec(const vector3 v3)
+{
+  switch (dim) {
+  case meep::D1:
+    return meep::vec(v3.z);
+  case meep::D2:
+    return meep::vec(v3.x, v3.y);
+  case meep::D3:
+    return meep::vec(v3.x, v3.y, v3.z);
+  case meep::Dcyl:
+    return meep::veccyl(v3.x, v3.z);
+  default:
+    meep::abort("unknown dimensionality in vector3_to_vec");
+  }
+}
+
+static geom_box gv2box(const meep::volume &v)
+{
+  geom_box box;
+  box.low = vec_to_vector3(v.get_min_corner());
+  box.high = vec_to_vector3(v.get_max_corner());
+  return box;
+}
+
+
+static bool is_variable(material_type mt)
+{
+  material_data *md = (material_data *)mt.data;
+  return (md->which_subclass == material_data::MATERIAL_FUNCTION);
+}
+
+static bool is_self(material_type mt)
+{
+  material_data *md = (material_data *)mt.data;
+  return (md->which_subclass == material_data::MATERIAL_TYPE_SELF);
+}
+
+static bool is_medium(material_type mt, medium_struct **m)
+{ 
+  material_data *md = (material_data *)mt.data;
+  if (md->which_subclass == material_data::MEDIUM)
+   { *m = md->medium;
+     return true;
+   };
+  return false;
+}
+
+static bool is_metal(meep::field_type ft, const material_type *material) {
+  material_data *md=(material_data *)material->data;
+  if (ft == meep::E_stuff)
+    switch (md->which_subclass) {
+    case material_data::MEDIUM:
+      return (md->medium->epsilon_diag.x < 0 ||
+              md->medium->epsilon_diag.y < 0 ||
+              md->medium->epsilon_diag.z < 0);
+    case material_data::PERFECT_METAL:
+      return true;
+    default:
+      meep::abort("unknown material type");
+  }
+  else
+    switch (md->which_subclass) {
+    case material_data::MEDIUM:
+      return (md->medium->mu_diag.x < 0 ||
+              md->medium->mu_diag.y < 0 ||
+              md->medium->mu_diag.z < 0);
+    case material_data::PERFECT_METAL:
+      return false; // is an electric conductor, but not a magnetic conductor
+    default:
+      meep::abort("unknown material type");
+  }
+}
+
+static meep::realnum *epsilon_data = NULL;
+static int epsilon_dims[3] = {0,0,0};
+
+void read_epsilon_file(const char *eps_input_file)
+{
+  delete[] epsilon_data; epsilon_data = NULL;
+  epsilon_dims[0] = epsilon_dims[1] = epsilon_dims[2] = 1;
+  if (eps_input_file && eps_input_file[0]) { // file specified
+    char *fname = new char[strlen(eps_input_file)+1];
```

Comment:
Where is this memory freed?  Do you think `std::string` would be safer and less error prone?